### PR TITLE
Fix segfault in AST optimization of `arrayExists` function

### PR DESCRIPTION
**Pull request by evillique**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix segmentation fault caused by AST optimization of `arrayExists` function applied in the JOIN ON part of the query.

cc: @taiyang-li

### CODE REVIEW

File: src/Interpreters/RewriteArrayExistsFunctionVisitor.cpp [hunk 1]
```diff
@@ -76,4 +77,22 @@ void RewriteArrayExistsFunctionMatcher::visit(const ASTFunction & func, ASTPtr &
     }
 }
 
+bool RewriteArrayExistsFunctionMatcher::needChildVisit(const ASTPtr & ast, const ASTPtr & child)
+{
+    /// SELECT ... JOIN ... ON arrayExists(x -> x = 1, arr)
```
Comment (alexey-milovidov) [inline src/Interpreters/RewriteArrayExistsFunctionVisitor.cpp:82]: The comment is not explanatory enough for such a specific code (why only `arrayExists`? smells badly)

File: tests/queries/0_stateless/02834_array_exists_segfault.sql [hunk 2]
```diff
@@ -0,0 +1,2 @@
+CREATE TABLE 02834_t (id UInt64, arr Array(UInt64)) ENGINE = MergeTree ORDER BY id;
+WITH subquery AS (SELECT []) SELECT t.* FROM 02834_t AS t JOIN subquery ON arrayExists(x -> x = 1, t.arr); -- { serverError INVALID_JOIN_ON_EXPRESSION }
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/02834_array_exists_segfault.sql:2]: Add DROP TABLE.

**Comment by robot-clickhouse-ci-2** [comment]
This is an automated comment for commit 4b764f7b6564c1723dd0547e377fd0c1c7dbddbd with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/54379/4b764f7b6564c1723dd0547e377fd0c1c7dbddbd/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Interpreters/RewriteArrayExistsFunctionVisitor.cpp (hunk 1 above)
Comment (evillique) [reply src/Interpreters/RewriteArrayExistsFunctionVisitor.cpp]: Changed the implementation to correctly apply the optimization even for incorrect JOIN ON expression.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by evillique** [comment]
Integration tests:
- https://github.com/ClickHouse/ClickHouse/issues/54465

**Pull request merged**